File: cpp/745_word_filter/word_filter.cpp
/*
 * @Date: 2022-07-14
 * @LastEditors: mengsenwang mengsen_wang@163.com
 * @LastEditTime: 2022-07-14
 * @FilePath: /algorithm/745_word_filter/word_filter.cpp
 */

#include <cassert>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class WordFilter {
 private:
  unordered_map<string, int> dict;

 public:
  WordFilter(vector<string> words) {
    for (int i = 0; i < words.size(); i++) {
      int m = words[i].size();
      string word = words[i];
      for (int prefixLength = 1; prefixLength <= m; prefixLength++) {
        for (int suffixLength = 1; suffixLength <= m; suffixLength++) {
          string key = word.substr(0, prefixLength) + '#' + word.substr(m - suffixLength);
          dict[key] = i;
        }
      }
    }
  }

  int f(string pref, string suff) {
    string target = pref + '#' + suff;
    return dict.count(target) ? dict[target] : -1;
  }
};

int main() {
  WordFilter w{vector<string>{"apple"}};
  assert(w.f("a", "e") == 0);
}

File: rust/745_word_filter/word_filter.rs
/*
 * @Date: 2022-07-14
 * @LastEditors: mengsenwang mengsen_wang@163.com
 * @LastEditTime: 2022-07-14
 * @FilePath: /algorithm/745_word_filter/word_filter.rs
 */

struct WordFilter {
    map: std::collections::HashMap<String, i32>,
}

impl WordFilter {
    fn new(words: Vec<String>) -> Self {
        let mut w = WordFilter {
            map: std::collections::HashMap::new(),
        };
        for i in 0..words.len() {
            let word = &words[i];
            let m = word.len();
            for prefix in 1..=m {
                for suffix in 1..=m {
                    let key = format!("{}{}{}", &word[0..prefix], "#", &word[m - suffix..m]);
                    w.map.insert(key, i as i32);
                }
            }
        }
        w
    }

    fn f(&self, pref: String, suff: String) -> i32 {
        let target: String = format!("{}{}{}", pref, "#", suff);
        if self.map.contains_key(&target) {
            *self.map.get(&target).unwrap()
        } else {
            -1
        }
    }
}

fn main() {
    let w = WordFilter::new(vec![String::from("apple")]);
    assert_eq!(w.f(String::from("a"), String::from("e")), 0);
}
